### DISCUSSION TITLE
Check permission set coverage

### DISCUSSION THREAD

**Discussion by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#88] [New Rule]
Add a rule that checks that each AL object is covered by at least one XML- (deprecated, I know, but still 😉) or AL-defined permission set.

I have already written a PowerShell script (see blog post section: [Checking Coverage of your Permission Sets](https://robvanbekkum.nl/additional-build-steps/#checking-coverage-of-your-permission-sets)) that can do this which can be added as a task in build pipelines, which could be used as the inspiration for a code analyzer. 😊
It would be a huge help to also see that some objects are not covered by any permission set in VS Code.

**Comment by fvet** [comment]
@rvanbekkum @StefanMaron 
In our ISV add-on, we tend to limit permission sets to **tabledata** level only (RIMD) and don't assign permissions on other object types (pages / codeunits / ...). Those objects types are all open (id **0**). Would it be possible to have a setting that indicates which objects to check permissions for? E.g. by default is it set to **All**, but with the option to change to (array of object types) e.g. **TableData, Pages, Reports, ....**

**Reply by StefanMaron** [reply]
I think that should be possible, yes

**Reply by jwikman** [reply]
Is it really necessary with a setting for this, shouldn't it be enough if the rule checked for wildcard object id's as well? I.e. Id `0` for the old Xml PermissionSet and object name `*` for the new PermissionSet object.

**Comment by StefanMaron** [comment]
Recently I had a longer discussion with KB about this rule. We agreed that this Rule is potentially rather often disabled.
Reason for this is, that you do not need any permissions besides `tabledata` if you rely on MS permission.
And the MS permission are needed to use BC in any way.

If some uses 100% custom permissions, they would also be responsible for covering the objects in installed extensions.

Because of this I will change this rule once again.
My Idea is to introduce a new setting in the `LinterCop.json` where you can specify which permission types will be checked.
By default the rule will then only check for `tabledata`. There is the exact same check in the PTE cop, but it seems like, for AppSource development, the PTECop is not always enabled.

**Reply by rvanbekkum** [reply]
> By default the rule will then only check for `tabledata`.

I don't agree with changing the current behavior. As far as I know, you can actually run into permission issues if you did not incorporate the permissions for the other object types - it's often not only `tabledata` permissions being necessary.

I think one could disable this rule and use AS0103 instead if one only wants to check table permissions.

**Reply by StefanMaron** [reply]
adding @mynavblog to the discussion

**Reply by jwikman** [reply]
I agree with @rvanbekkum. AS0103 now exists for AppSourceCop, it checks for tabledata permissions. (works just as PTE0004).

Having this rule to also only check for tabledata permissions makes it a bit unnecessary. 

So leave the rule as is, and then you can disable the rule if you do not want the additional permission checks.
In our AppSource apps we want to include all execute permissions, for those cases where the customer has a tight permission structure - and do not use MS Default permission sets.

**Comment by mynavblog** [comment]
@rvanbekkum Permission for other objects are granted in the standard permission set - D365 Basic. There is no requirement from MS to add other permission to your custom apps. Also MS permission sets contain only **tabledata** permissions. This is to simplify the process of permissions. If you would not add permission to D365 Basic no standard permission would work. Which also means someone is doing big redesign of permission sets anyway. 

Therefore in 98% of cases this rule would be mostly disabled.

@jwikman I agree that in that case it works almost the same as PTE0004 but this is in PTE Cop and what is missed for many partners is the same rule in **AppSourceCop**. Therefore the setup would be great - it is much simple than turning on the PTE Cop and then disable all the rules except this one.

**Reply by jwikman** [reply]
As Rob mentioned, there is now a rule in AppSourceCop (AS0103, if I recall..) that works just as PTE0004. So making this rule work just as those two rules doesn't make sense for me. 🙂

**Reply by StefanMaron** [reply]
@jwikman Agree. If there is a rule in the appsource cop now, there is no need to change this rule. 

Then, the only question is, if it is a good idea to "force" this rule on everyone. Or if it would be better to have this one disabled by default, and everyone who wants to use it, can enable it.

**Reply by jwikman** [reply]
I would be fine if this rule became "opt-in" rather than "opt-out". 👍

**Reply by mynavblog** [reply]
I think it is new rule since it is not listed yet on the list https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/analyzers/appsourcecop  - if so then I do not see the value to force most of the people to disable the LinterCop at start. 

In most projects tabledata is enough.

**Reply by fvet** [reply]
Where did you guys found out about rule AS0103? :)

**Reply by jwikman** [reply]
> Where did you guys found out about rule AS0103? :)

https://github.com/microsoft/AL/issues/6503#issuecomment-1055576571 🙂
